File: ReLeetCode/0021.py
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        l1 = list1
        l2 = list2
        cur_ptr = dummy

        while l1 and l2:
            if l1.val < l2.val:
                cur_ptr.next = l1
                l1 = l1.next
                cur_ptr = cur_ptr.next
            else:
                cur_ptr.next = l2
                l2 = l2.next
                cur_ptr = cur_ptr.next

        while l1:
            cur_ptr.next = l1
            l1 = l1.next
            cur_ptr = cur_ptr.next
        
        while l2:
            cur_ptr.next = l2
            l2 = l2.next
            cur_ptr = cur_ptr.next

        return dummy.next

# 申请一个dummy node来简化合并操作

# 时间复杂度 O(m+n)
# 空间复杂度 O(1)

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if list1 == None and list2 == None:
            return None
        if list1 != None and list2 == None:
            return list1
        if list2 != None and list1 == None:
            return list2
        if list1.val <= list2.val:
            list1.next = self.mergeTwoLists(list1.next, list2)
            return list1
        else:
            list2.next = self.mergeTwoLists(list1, list2.next)
            return list2


# 使用递归来实现, 递归的出口为两个list都为空,或者其中一个为空
# 如果list1.val <= list2.val, 说明list1为合并后链表的第一个节点, list1.next为list1.next和list2合并之后的结果
# 最后返回list1
# 如果list1.val > list2.val, 与前面的相反
# 最后返回list2

# 时间复杂度 O(m + n)
# 空间复杂度 O(m + n)

File: CppLeetCode/0021.cpp
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if (list1 == nullptr && list2 == nullptr) {
            return nullptr;
        }
        if (list1 == nullptr && list2 != nullptr) {
            return list2;
        }
        if (list1 != nullptr && list2 == nullptr) {
            return list1;
        }
        if (list1->val < list2->val) {
            list1->next = mergeTwoLists(list1->next, list2);
            return list1;
        } 
        list2->next = mergeTwoLists(list2->next, list1);
        return list2;
    }
};

// 时间复杂度 O(n+m) 空间复杂度 O(n+n) 递归法

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if (list1 == nullptr && list2 == nullptr) {
            return nullptr;
        }
        if (list1 == nullptr && list2 != nullptr) {
            return list2;
        }
        if (list1 != nullptr && list2 == nullptr) {
            return list1;
        }
        ListNode* dummy_node = new ListNode(-1);
        ListNode* pre = dummy_node;
        ListNode* p1 = list1;
        ListNode* p2 = list2;
        while (p1 != nullptr && p2 != nullptr) {
            if (p1->val < p2->val) {
                dummy_node->next = p1;
                p1 = p1->next;
            } else {
                dummy_node->next = p2;
                p2 = p2->next;
            }
            dummy_node = dummy_node->next;
        }
        if (p1 != nullptr) {
            dummy_node->next = p1;
        } else {
            dummy_node->next = p2;
        }
        return pre->next;
    }
};

// 时间复杂度 O(n+m) 空间复杂度 O(1) 迭代法